src/main/rx/fport.c

```diff
@@ -147,57 +147,54 @@ static void fportDataReceive(uint16_t c, void *data)
     UNUSED(data);
 
     static timeUs_t frameStartAt = 0;
-    static bool frameStarting = false;
     static bool escapedCharacter = false;
-    static uint8_t bufferPosition = 0;
     static timeUs_t lastFrameReceivedUs = 0;
     static bool telemetryFrame = false;
 
     const timeUs_t currentTimeUs = micros();
 
-    if (!skipUntilStart && (frameStarting || bufferPosition > 0) && cmpTimeUs(currentTimeUs, frameStartAt) > FPORT_TIME_NEEDED_PER_FRAME_US + 500) {
-         reportFrameError(DEBUG_FPORT_ERROR_TIMEOUT);
+    if (framePosition > 0) {
+        clearToSend = false;
 
-         bufferPosition = 0;
-         clearToSend = false;
+        if (cmpTimeUs(currentTimeUs, frameStartAt) > FPORT_TIME_NEEDED_PER_FRAME_US + 500) {
+            reportFrameError(DEBUG_FPORT_ERROR_TIMEOUT);
+
+            framePosition = 0;
+        }
      }
 
     uint8_t val = (uint8_t)c;
 
     if (val == FPORT_FRAME_MARKER) {
-        skipUntilStart = false;
-        frameStarting = true;
-        escapedCharacter = false;
-        frameStartAt = currentTimeUs;
-
-        if (bufferPosition > 0) {
-            uint8_t nextWriteIndex = (rxBufferWriteIndex + 1) % NUM_RX_BUFFERS;
+        if (framePosition > 0) {
```

Comment:
There is one special case that may need handling:
If ending `FPORT_FRAME_MARKER` is seen first, it will set `framePosition = 1`. Then starting `FPORT_FRAME_MARKER` will terminate zero-length packet. Following frame content would be discarded as inter-frame garbage. Without timeout, this will repeat forever.

Timeout will step in and force resync (if space between packet is larger that FPORT_TIME_NEEDED_PER_FRAME_US), but it can be handled explicitly - assume that double marker is start of new packet.

`if (framePosition > 1) {` should be all that is necessary; special-case for DEBUG_SET is possible.

Reply:
Ok, let's turn this onto its head, and assume that the frame end marker might or might not get lost. I.e. we are treating every marker now as if it might be a frame end, and only sending out bytes will set `framePosition = 0` to make the state machine wait until the next marker.

Reply:
Yes, implementation in last commit looks great, I don't see any problem. Thanks for your patience

Reply:
Thank you for helping with this.